clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -384,34 +384,50 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 return false;
             }
 
-            // call onJoinPrepare if needed. We set a flag to make sure that we do not call it a second
-            // time if the client is woken up before a pending rebalance completes. This must be called
-            // on each iteration of the loop because an event requiring a rebalance (such as a metadata
-            // refresh which changes the matched subscription set) can occur while another rebalance is
-            // still in progress.
-            if (needsJoinPrepare) {
-                // need to set the flag before calling onJoinPrepare since the user callback may throw
-                // exception, in which case upon retry we should not retry onJoinPrepare either.
-                needsJoinPrepare = false;
-                onJoinPrepare(generation.generationId, generation.memberId);
+            RequestFuture<ByteBuffer> future;
+
+            synchronized (this) {
+                // call onJoinPrepare if needed. We set a flag to make sure that we do not call it a second
+                // time if the client is woken up before a pending rebalance completes. This must be called
+                // on each iteration of the loop because an event requiring a rebalance (such as a metadata
+                // refresh which changes the matched subscription set) can occur while another rebalance is
+                // still in progress.
+                if (needsJoinPrepare) {
+                    // need to set the flag before calling onJoinPrepare since the user callback may throw
+                    // exception, in which case upon retry we should not retry onJoinPrepare either.
+                    needsJoinPrepare = false;
+                    onJoinPrepare(generation.generationId, generation.memberId);
+                }
+
+                future = initiateJoinGroup();
             }
 
-            final RequestFuture<ByteBuffer> future = initiateJoinGroup();
             client.poll(future, timer);
             if (!future.isDone()) {
                 // we ran out of time
                 return false;
             }
 
             if (future.succeeded()) {
-                // Duplicate the buffer in case `onJoinComplete` does not complete and needs to be retried.
-                ByteBuffer memberAssignment = future.value().duplicate();
-                onJoinComplete(generation.generationId, generation.memberId, generation.protocol, memberAssignment);
+                synchronized (this) {
+                    // Generation data maybe concurrently cleared by Heartbeat thread.
+                    if (generation.generationId != OffsetCommitRequest.DEFAULT_GENERATION_ID) {
```

Comment:
We can just compare the whole generation object with Generation.NO_GENERATION.

Reply:
fixed.